ethylene-core/src/main/java/com/github/steanky/ethylene/core/processor/FileConfigLoader.java

```diff
@@ -0,0 +1,103 @@
+package com.github.steanky.ethylene.core.processor;
+
+import com.github.steanky.ethylene.core.ConfigElement;
+import com.github.steanky.ethylene.core.bridge.ConfigBridge;
+import com.github.steanky.ethylene.core.codec.ConfigCodec;
+import com.github.steanky.ethylene.core.util.FutureUtils;
+import org.jetbrains.annotations.NotNull;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Objects;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Filesystem-source specialization of ProcessingConfigLoader. The source is considered <i>absent</i> if the given
+ * {@link Path} does not exist (see {@link ProcessingConfigLoader} for more information on how an absent source changes
+ * behavior).
+ * @param <TData> the type of data object
+ */
+public class FileConfigLoader<TData> extends ProcessingConfigLoader<TData> {
+    /**
+     * The {@link Path} object acting as the source of data.
+     */
+    protected final Path path;
+
+    private final ConfigBridge bridge;
+
+    /**
+     * General ConfigBridge implementation which may be synchronous or asynchronous. It will encode and decode data
+     * based off of a single, provided codec.
+     */
+    protected abstract class CodecConfigBridge implements ConfigBridge {
+        private final ConfigCodec codec;
+
+        /**
+         * Produces a new instance of CodecConfigBridge using the provided codec.
+         * @param codec the codec to use
+         */
+        protected CodecConfigBridge(@NotNull ConfigCodec codec) {
+            this.codec = Objects.requireNonNull(codec);
+        }
+
+        @Override
+        public @NotNull CompletableFuture<ConfigElement> read() {
+            return makeFuture(() -> {
+                codec.decode(Files.newInputStream(path));
+                return null;
```

Comment:
this looks sus unless I'm misinterpreting how this works

Reply:
that is indeed sus, fixed

Reply:
(i'd like to unit test all of these but unit testing filesystem stuff is super cringe)

Reply:
we might need to instead pass input streams around rather than Paths, bit annoying but meh

Reply:
for now this is fine, i don't feel like butchering my API just to make way for tests